### PR TITLE
[Enhencement](Nereids) reduce child output rows if agg child is literal

### PR DESCRIPTION
**Pull request by LiBinfeng-01**
## Proposed changes

with group by:
select max(1) from t1 group by c1; -> select 1 from (select c1 from t1 group by c1);
without group by:
select max(1) from t1; -> select max(1) from (select 1 from t1 limit 1) tmp;

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run feut

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalAggregate.java [hunk 1]
```diff
@@ -83,6 +83,8 @@ public class LogicalAggregate<CHILD_TYPE extends Plan>
     private final boolean generated;
     private final boolean hasPushed;
 
+    private boolean isRewritten = false;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalAggregate.java:86]: the variable name is too generic

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateAggregateConstant.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+                    }
+                } else {
+                    return ne;
+                }
+            }).collect(Collectors.toList());
+
+            if (agg.getGroupByExpressions().isEmpty()) {
+                LogicalAggregate newAgg = new LogicalAggregate<>(agg.getGroupByExpressions(),
+                        agg.getOutputExpressions(), new LogicalLimit(1, 0, LimitPhase.ORIGIN,
+                                new LogicalProject<>(newOutput, agg.child())));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateAggregateConstant.java:71]: this lead to project output and agg output has two slot with same exprid, it is not right.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateAggregateConstant.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                newAgg.setRewritten(true);
+                return newAgg;
+            } else {
+                List<NamedExpression> childOutput = agg.getOutputExpressions().stream().filter(ne ->
+                        !(ne instanceof Alias && ne.child(0) instanceof AggregateFunction
+                            && (ne.child(0) instanceof Max || ne.child(0) instanceof Min)))
+                            .collect(Collectors.toList());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateAggregateConstant.java:78]: hard to understand, we only put all group by key to output is easy for code and understand

**Comment by LiBinfeng-01** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ReduceAggregateChildOutputRows.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        return logicalAggregate().then(agg -> {
+            Set<AggregateFunction> aggFunctions = agg.getAggregateFunctions();
+            // check whether we have aggregate(constant) in all aggregateFunctions
+            if (!agg.isRewrittenByAggregateConstant() || aggFunctions.isEmpty() || !aggFunctions.stream().allMatch(
+                    f -> (f instanceof Min || f instanceof Max)
+                            && (f.arity() == 1 && f.child(0).isConstant()))) {
+                return null;
+            }
+
+            List<NamedExpression> newOutput = agg.getOutputExpressions().stream().map(ne -> {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ReduceAggregateChildOutputRows.java:55]: avoid to use stream api for better performance

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ReduceAggregateChildOutputRows.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                    AggregateFunction f = (AggregateFunction) ne.child(0);
+                    if (f instanceof Min || f instanceof Max) {
+                        return new Alias(ne.getExprId(), f.child(0), ne.getName());
+                    } else {
+                        throw new AnalysisException("Unexpected aggregate function: " + f);
+                    }
+                } else {
+                    return ne;
+                }
+            }).collect(Collectors.toList());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ReduceAggregateChildOutputRows.java:66]: use ImmutableList Builder to reduce list copy

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/RuleType.java [hunk 6]
```diff
@@ -199,6 +199,8 @@ public enum RuleType {
     // Eliminate plan
     MERGE_AGGREGATE(RuleTypeClass.REWRITE),
     ELIMINATE_AGGREGATE(RuleTypeClass.REWRITE),
+
+    ELIMINATE_AGGREGATE_SUM(RuleTypeClass.REWRITE),
```
[the remark this answers is not in the document]
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/RuleType.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateAggregateConstant.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+import org.apache.doris.nereids.util.PlanUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/** EliminateAggregateSum
+ * with group by:
+ * select max(1) from t1 group by c1; -> select 1 from t1;
```
[the remark this answers is not in the document]
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateAggregateConstant.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateAggregateConstant.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+ * without group by:
+ * select max(1) from t1; -> select max(1) from (select 1 from t1 limit 1) tmp;
+ *
+ * extend:
+ * select max(1) + min(c1) from t1 group by c1;
```
[the remark this answers is not in the document]
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateAggregateConstant.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateAggregateConstant.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+            List<String> columnAliases = new ArrayList<>();
+            List<NamedExpression> newOutput = agg.getOutputExpressions().stream().map(ne -> {
+                if (ne instanceof Alias && ne.child(0) instanceof AggregateFunction) {
+                    AggregateFunction f = (AggregateFunction) ne.child(0);
+                    if (f instanceof Sum || f instanceof Min || f instanceof Max) {
+                        columnAliases.add(ne.getName());
+                        return new Alias(ne.getExprId(), f.child(0), ne.getName());
+                    } else {
+                        throw new IllegalStateException("Unexpected aggregate function: " + f);
```
[the remark this answers is not in the document]
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateAggregateConstant.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/algebra/Aggregate.java [hunk 10]
```diff
@@ -53,6 +53,10 @@ default Set<AggregateFunction> getAggregateFunctions() {
         return ExpressionUtils.collect(getOutputExpressions(), AggregateFunction.class::isInstance);
     }
 
+    default Set<AggregateFunction> getAggregateFunctionsConst() {
+        return ExpressionUtils.collect(getOutputExpressions(), AggregateFunction.class::isInstance);
+    }
```
[the remark this answers is not in the document]
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/algebra/Aggregate.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateAggregateConstant.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+ *
+ * extend:
+ * select max(1) + min(c1) from t1 group by c1;
+ * */
+public class EliminateAggregateConstant extends OneRewriteRuleFactory {
+    @Override
+    public Rule build() {
+        return logicalAggregate().then(agg -> {
+            // check whether we have aggregate(constant) in all aggregateFunctions
+            Set<AggregateFunction> aggFunctions = agg.getAggregateFunctions();
```
[the remark this answers is not in the document]
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateAggregateConstant.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateAggregateConstant.java [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/** EliminateAggregateSum
+ * with group by:
+ * select max(1) from t1 group by c1; -> select 1 from t1 group by c1;
+ * without group by:
+ * select max(1) from t1; -> select max(1) from (select 1 from t1 limit 1) tmp;
+ * */
+public class EliminateAggregateConstant extends OneRewriteRuleFactory {
```
[the remark this answers is not in the document]
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateAggregateConstant.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalAggregate.java (hunk 1 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalAggregate.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateAggregateConstant.java (hunk 2 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateAggregateConstant.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateAggregateConstant.java (hunk 3 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateAggregateConstant.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ReduceAggregateChildOutputRows.java (hunk 4 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ReduceAggregateChildOutputRows.java]: changed to immutable list

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ReduceAggregateChildOutputRows.java (hunk 5 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ReduceAggregateChildOutputRows.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalAggregate.java [hunk 13]
```diff
@@ -83,6 +83,8 @@ public class LogicalAggregate<CHILD_TYPE extends Plan>
     private final boolean generated;
     private final boolean hasPushed;
 
+    private boolean isRewrittenByAggregateConstant = false;
```
[the remark this answers is not in the document]
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalAggregate.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/RuleType.java [hunk 14]
```diff
@@ -201,6 +201,8 @@ public enum RuleType {
     // Eliminate plan
     MERGE_AGGREGATE(RuleTypeClass.REWRITE),
     ELIMINATE_AGGREGATE(RuleTypeClass.REWRITE),
+
+    ELIMINATE_AGGREGATE_CONSTANT(RuleTypeClass.REWRITE),
```
[the remark this answers is not in the document]
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/RuleType.java]: done

**Comment by LiBinfeng-01** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ReduceAggregateChildOutputRows.java [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+
+            if (agg.getGroupByExpressions().isEmpty()) {
+                LogicalAggregate newAgg = new LogicalAggregate<>(agg.getGroupByExpressions(),
+                        agg.getOutputExpressions(), new LogicalLimit(1, 0, LimitPhase.ORIGIN,
+                                new LogicalProject<>(newOutput.build(), agg.child())));
+                newAgg.setRewrittenByReduceAggregateChildOutputRows(true);
+                return newAgg;
+            } else {
+                List<NamedExpression> childOutput = agg.getGroupByExpressions().stream().map(expr ->
+                        (NamedExpression) expr).collect(Collectors.toList());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ReduceAggregateChildOutputRows.java:81]: avoid stream api, use ImmutableList. and maybe use remove generic trick could avoid generate new list here?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ReduceAggregateChildOutputRows.java [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+ * without group by:
+ * select max(1) from t1; -> select max(1) from (select 1 from t1 limit 1) tmp;
+ * */
+public class ReduceAggregateChildOutputRows extends OneRewriteRuleFactory {
+    @Override
+    public Rule build() {
+        return logicalAggregate().then(agg -> {
+            Set<AggregateFunction> aggFunctions = agg.getAggregateFunctions();
+            // check whether we have aggregate(constant) in all aggregateFunctions
+            if (!agg.isRewrittenByReduceAggregateChildOutputRows()
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ReduceAggregateChildOutputRows.java:51]: could we avoid to use this flag by check whether this agg is scalar agg with only min max agg function and with limit 1 child?

**Comment by LiBinfeng-01** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ReduceAggregateChildOutputRows.java [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+ * without group by:
+ * select max(1) from t1; -> select max(1) from (select 1 from t1 limit 1) tmp;
+ * */
+public class ReduceAggregateChildOutputRows extends OneRewriteRuleFactory {
+    @Override
+    public Rule build() {
+        return logicalAggregate().then(agg -> {
+            Set<AggregateFunction> aggFunctions = agg.getAggregateFunctions();
+            // check whether we have aggregate(constant) in all aggregateFunctions
+            if (!(agg.child() instanceof LogicalLimit)
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ReduceAggregateChildOutputRows.java:50]: better to check limit is 1?

**Comment by LiBinfeng-01** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ReduceAggregateChildOutputRows.java [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+import org.apache.doris.nereids.trees.plans.logical.LogicalLimit;
+import org.apache.doris.nereids.trees.plans.logical.LogicalProject;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.Set;
+
+/** ReduceAggregateChildOutputRows
+ * with group by:
+ * select max(1) from t1 group by c1; -> select 1 from (select c1 from t1 group by c1);
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ReduceAggregateChildOutputRows.java:40]: why wrap it with a view？Does 'select 1 from t1 group by c1' ok?

**Review by xzj7019 — APPROVED** [review]

**Pull request merged**